crates/circuit/src/circuit_instruction.rs

```diff
@@ -63,15 +63,37 @@ pub struct CircuitInstruction {
     pub clbits: Py<PyTuple>,
 }
 
+impl CircuitInstruction {
+    pub fn new<T1, T2, U1, U2>(
+        py: Python,
+        operation: PyObject,
+        qubits: impl IntoIterator<Item = T1, IntoIter = U1>,
+        clbits: impl IntoIterator<Item = T2, IntoIter = U2>,
+    ) -> PyResult<Py<Self>>
+    where
+        T1: ToPyObject,
+        T2: ToPyObject,
+        U1: ExactSizeIterator<Item = T1>,
+        U2: ExactSizeIterator<Item = T2>,
+    {
+        CircuitInstruction::py_new(
+            py,
+            operation,
+            Some(&PyTuple::new_bound(py, qubits)),
+            Some(&PyTuple::new_bound(py, clbits)),
+        )
+    }
+}
+
 #[pymethods]
 impl CircuitInstruction {
     #[new]
-    pub fn new(
+    pub fn py_new(
```

Comment:
Oh, I like this pattern, this is something we need to do a lot and have a dedicated python space constructor like this works well.